### ISSUE TITLE
Add 'style' env to Travis to add hlint-ing for stack code

### ISSUE THREAD

**Issue by lwm** [commercialhaskell/stack#2592]
Following from https://github.com/commercialhaskell/stack/pull/2591#issuecomment-246676805, as recommended by @Blaisorblade.

I'd like to see if there is consensus amongst contributors to include hlint as a quality assurance tool for stack. There are currently 355 hints to deal with when running:

```
  $ find src -name "*hs" -exec hlint {} + ;
```

If people think this is a good change (IMHO, it is ;)), then I will do the following:

  - [x] Remove all warnings (let's see how it goes)
    - #2606, #2615, #2690
  - [x] Document a low overhead way to work with hlint and the stack codebase
    - https://github.com/commercialhaskell/stack/pull/2699
  - [x] Rebase #2591 and remove `allowed_failures` config

**Comment by Blaisorblade** [comment]
Here's my 👍 , especially since you volunteer to fix this. If/when you get started on warnings, please make sure to separate potential behavioral changes, like ones that alter laziness:

```
src/Data/Attoparsec/Interpreter.hs:150:14: Suggestion: Use null
Found:
  length ctxs == 0
Why not:
  null ctxs
Note: increases laziness
```

**Comment by Blaisorblade** [comment]
Pinging @sjakobi, @mgsloan, @harendra-kumar and @borsboom for opinions on addressing `hlint` warnings — @lwm is volunteering to do the work and already started in #2588.

In case there's no objection for a while, you might want to get started. As far as I can tell, we usually don't wait for explicit unanymous agreement for every little detail 😉 
https://github.com/commercialhaskell/stack/pull/2591#issuecomment-246679510

Also, googling suggests there's not so much controversy over hlint itself—people debate warning categories though. [Quoting hlint's author](http://neilmitchell.blogspot.de/2009/09/how-i-use-hlint.html):

> I do not intend people to slavishly apply the hints given by HLint - each hint should be carefully considered. For example, the darcs project uses HLint, but has decided that they are not interested in eta reduction hints, so have used HLint's ignoring facility.

**Comment by lwm** [comment]
@Blaisorblade 

Thanks for the link. It's helpful. As with https://github.com/commercialhaskell/stack/pull/2606, I will move through each type of suggestion/warning and see what I can apply programmatically (`hlint --refactor --refactor-options -i`) and what I can reason about and manually do. Anything that seems questionable, we can start to use the ignore facility.

**Comment by Blaisorblade** [comment]
Fixed by #2591 I guess.
